diff --git a/node_modules/urkel/lib/store.js b/node_modules/urkel/lib/store.js
index 17d8b79..1081ac2 100644
--- a/node_modules/urkel/lib/store.js
+++ b/node_modules/urkel/lib/store.js
@@ -601,10 +601,13 @@ class Store {
     state.rootPtr = root.ptr;
     state.rootNode = root.toHash(this.hash);
 
-    const padding = META_SIZE - (this.buffer.pos % META_SIZE);
+    if (this.buffer.pos % META_SIZE) {
+      const padding = META_SIZE - (this.buffer.pos % META_SIZE);
+      this.buffer.expand(padding);
+      this.buffer.pad(padding);
+    }
 
-    this.buffer.expand(padding + META_SIZE);
-    this.buffer.pad(padding);
+    this.buffer.expand(META_SIZE);
 
     const off = state.write(
       this.buffer.data,
